client/data/followers/use-followers-query.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * External dependencies
+ */
+import { useInfiniteQuery } from 'react-query';
+import { uniqueBy } from '@automattic/js-utils';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'calypso/lib/wp';
+
+const MAX_FOLLOWERS = 10; // means pages (= 1000 followers);
+const defaults = {
+	max: 100,
+};
+
+/**
+ * Normalizes a follower object. Changes 'avatar' to 'avatar_URL' allowing a follower
+ * object to be used with the Gravatar component.
+ *
+ * @param  {object} follower A follower ojbect
+ * @returns {object}          A normalized follower object
+ */
+export function normalizeFollower( follower ) {
+	follower.avatar_URL = follower.avatar;
+	return follower;
+}
+
+const extractPages = ( pages = [] ) =>
+	pages.flatMap( ( page ) => page.subscribers ).map( normalizeFollower );
+const compareUnique = ( a, b ) => a.ID === b.ID;
+
+const useFollowersQuery = ( siteId, type = 'wpcom', fetchOptions = {}, queryOptions = {} ) => {
+	const { search } = fetchOptions;
+
+	return useInfiniteQuery(
+		[ 'followers', siteId, type, search ],
+		async ( { pageParam = 1 } ) =>
+			wpcom.req.get( `/sites/${ siteId }/followers`, {
+				...defaults,
+				...fetchOptions,
+				type,
+				page: pageParam,
+			} ),
+		{
+			...queryOptions,
+			getNextPageParam: ( lastPage, allPages ) => {
+				if ( lastPage.pages <= allPages.length || allPages.length >= MAX_FOLLOWERS ) {
+					return;
+				}
+				return allPages.length + 1;
+			},
+			select: ( data ) => {
+				return {
+					/* @TODO: `uniqueBy` is necessary, because the API can return duplicates */
+					followers: uniqueBy( extractPages( data.pages ), compareUnique ),
+					total: data.pages[ 0 ].total,
+					...data,
```

Comment:
My intention was not to hide anything which is "natively" returned by the [`useInfiniteQuery`](https://react-query.tanstack.com/reference/useInfiniteQuery) hook to ensure we're not hiding any documented props from consumers. It's debatable whether that approach is how we want to use custom query hooks and probably part of us developing conventions using ReactQuery.

/cc @jsnajdr might have an opinion here as well.

Reply:
By the way I found [this note in the ReactQuery](https://react-query.tanstack.com/guides/infinite-queries) docs which says

> When using options like `initialData` or `select` in your query, make sure when you restructure your data that it still includes `data.pages` and `data.pageParams` properties, otherwise your changes will be overwritten by the query in its return

Reply:
I don't know yet what is going to be the ideal API for hooks built on `useQuery` or `useMutation`. Should we return only some selected data and hide everything else from the low-level hooks? Or should we only add new properties and expose the underlying `react-query` API fully? We'll see as we start using the library at more places.